pallets/foreign-investments/src/impls/invest.rs

```diff
@@ -1085,3 +1085,279 @@ where
+					..pool_swap
+				}
+			}
+		);
+
+		assert_ok!(
+			InvestState::ActiveSwapIntoForeignCurrency { swap: foreign_swap }
+				.transition(increase.clone()),
+			InvestState::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDoneAndInvestmentOngoing {
+				swap: Swap {
+					amount: foreign_swap.amount - to_foreign(pool_swap.amount),
+					..foreign_swap
+				},
+				done_amount: 40,
+				invest_amount: pool_swap.amount,
+			}
+		);
+
+		assert_ok!(
+			InvestState::ActiveSwapIntoPoolCurrencyAndInvestmentOngoing {
+				swap: pool_swap,
+				invest_amount
+			}
+			.transition(increase.clone()),
+			InvestState::ActiveSwapIntoPoolCurrencyAndInvestmentOngoing {
+				swap: Swap {
+					amount: pool_swap.amount + pool_swap.amount,
+					..pool_swap
+				},
+				invest_amount
+			}
+		);
+
+		assert_ok!(
+			InvestState::ActiveSwapIntoForeignCurrencyAndInvestmentOngoing {
+				swap: foreign_swap,
+				invest_amount
+			}
+			.transition(increase.clone()),
+			InvestState::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDoneAndInvestmentOngoing {
+				swap: Swap {
+					amount: foreign_swap.amount - to_foreign(pool_swap.amount),
+					..foreign_swap
+				},
+				done_amount: 40,
+				invest_amount: invest_amount + pool_swap.amount,
+			}
+		);
+
+		assert_err!(
+			InvestState::ActiveSwapIntoPoolCurrencyAndSwapIntoForeignDone {
+				swap: pool_swap,
+				done_amount,
+			}
+			.transition(increase.clone()),
+			DispatchError::Other(
+				"Invalid invest state, should automatically be transitioned into \
+				 ActiveSwapIntoPoolCurrencyAndInvestmentOngoing",
+			)
+		);
+
+		assert_ok!(
+			InvestState::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDone {
+				swap: foreign_swap,
+				done_amount,
+			}
+			.transition(increase.clone()),
+			InvestState::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDoneAndInvestmentOngoing {
+				swap: Swap {
+					amount: foreign_swap.amount - to_foreign(pool_swap.amount),
+					..foreign_swap
+				},
+				done_amount: 100,
+				invest_amount: pool_swap.amount
+			}
+		);
+
+		assert_ok!(
+			InvestState::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDoneAndInvestmentOngoing {
+				swap: foreign_swap,
+				done_amount,
+				invest_amount,
+			}
+			.transition(increase.clone()),
+			InvestState::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDoneAndInvestmentOngoing {
+				swap: Swap {
+					amount: foreign_swap.amount - to_foreign(pool_swap.amount),
+					..foreign_swap
+				},
+				done_amount: 100,
+				invest_amount: invest_amount + pool_swap.amount
+			}
+		);
+	}
+
+	struct Checker; // TODO: add a checker interal state
+
+	impl Checker {
+		fn feed_with_step(&mut self, transition: &InvestTransition) {
+			// TODO: mutate checker state to track the happen transition into
+			// something with meaning for the invariant computation
+		}
+
+		fn check_invariant(&self, state: &InvestState) {
+			// TODO: evaluate the invariant for the current state and check with
+			// an assert
+		}
```

Comment:
@wischli could you help me by suggesting a good invariant here? Or with some basic rule I could compute from a list of `transactions` and comparable to the final `state` after applying those transactions?